### PR TITLE
Backfill scholarship course

### PR DESCRIPTION
**Pull request by clareconstantine**
Another step in [PLC-165](https://codedotorg.atlassian.net/browse/PLC-165)

This script fills in the newly added `course`column on the `ScholarshipInfo` table. If the user has an accepted `Teacher1920Application`, we use the course from the application. If they have an enrollment in a local summer workshop for this year, we check to make sure they aren't a new facilitator (with an accepted `Facilitator1920Application`). If they are, we don't update that `ScholarshipInfo` (because they shouldn't have one) and print a message to the console. If they are not, then they are a teacher whose partner did not use our application system and we use the course from their enrollment.

### CODE REVIEW

**Comment by breville** [comment]
Two questions about the description:
- How does somebody end up in a summer workshop but are also a facilitator?  Is this a typical occurrence?
- If they are in a summer workshop but are not a facilitator, how do we know they are a teacher whose partner did not use our application system?

File: bin/oneoff/backfill_data/backfill_scholarship_course [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+def course_from_current_teacher_accepted_app(si)
+  app = Pd::Application::Teacher1920Application.find_by(user_id: si.user_id)
+  if app && Pd::SharedApplicationConstants::COHORT_CALCULATOR_STATUSES.include?(app.status)
+    return app.course
+  end
+  nil
+end
+
+def course_from_current_teacher_lsw_enrollment(si)
```
Comment (breville) [inline bin/oneoff/backfill_data/backfill_scholarship_course:19]: what is `lsw`?

File: bin/oneoff/backfill_data/backfill_scholarship_course [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  enrollments = Pd::Enrollment.where(user_id: si.user_id)
+  course = nil
+  enrollments.each do |e|
+    next unless e.workshop.local_summer? && e.workshop.summer_workshop_school_year == "2019-2020"
```
Comment (breville) [inline bin/oneoff/backfill_data/backfill_scholarship_course]: nit: is there a constant we can use for this string?

File: bin/oneoff/backfill_data/backfill_scholarship_course [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  course
+end
+
+ActiveRecord::Base.transaction do
+  count = 0
+  Pd::ScholarshipInfo.find_each do |si|
+    course = course_from_current_teacher_accepted_app(si) || course_from_current_teacher_lsw_enrollment(si)
+    if course
+      puts "#{si.id} #{course}"
+      # This script is a dry run unless we comment out this update line
```
Comment (breville) [inline bin/oneoff/backfill_data/backfill_scholarship_course]: I like the dry run approach!

File: bin/oneoff/backfill_data/backfill_scholarship_course (hunk 1 above)
Comment (hacodeorg) [reply bin/oneoff/backfill_data/backfill_scholarship_course:19]: may be local summer workshop?

**Comment by clareconstantine** [comment]
@breville, the new facilitators, people who have applied and been accepted as facilitators this year, attend local summer workshops to observe and learn how to facilitate. I asked SarahFair, and there are no new facilitators coming through partners who didn’t use our application

**Comment by hacodeorg** [comment]
What is the result of your dry run? How many rows in ScholarshipInfo and how many will be updated?

**Comment by clareconstantine** [comment]
@hacodeorg I haven't run it yet because i didn't want to run it on production without a review. Once this is in production, I'll do the dry run to verify the outcome is what i would expect and then run it for real.

**Comment by clareconstantine** [comment]
@hacodeorg but as of yesterday there were 2317 rows in production, and one was for a facilitator. so i would expect all but that row to be updated.

File: bin/oneoff/backfill_data/backfill_scholarship_course [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+  enrollments = Pd::Enrollment.where(user_id: si.user_id)
+  course = nil
+  enrollments.each do |e|
+    next unless e.workshop.local_summer? && e.workshop.summer_workshop_school_year == Pd::Application::ActiveApplicationModels::APPLICATION_CURRENT_YEAR
+
+    if Pd::Application::Facilitator1920Application.find_by(user_id: si.user_id)&.status == 'accepted'
+      puts "user #{si.user_id} new facilitator, app #{Pd::Application::Facilitator1920Application.find_by(user_id: si.user_id).id}"
+    end
+    if e.workshop.course == "CS Discoveries"
+      course = 'csd'
```
Comment (breville) [inline bin/oneoff/backfill_data/backfill_scholarship_course:29]: Should we break out of the loop here?

Comment (breville) [reply bin/oneoff/backfill_data/backfill_scholarship_course:29]: Or notice when we set `course` more than once?

**Review by breville — APPROVED** [review]

**Comment by breville** [comment]
Thanks for walking me through this!  I hadn't really internalised that some teachers signing up for a workshop have filled out a teacher application first, while others haven't.

**Comment by hacodeorg** [comment]
Thanks. I'd suggest for the database updating code like this we should run a few queries to get a clear expectation of the input and output.

> @hacodeorg but as of yesterday there were 2317 rows in production, and one was for a facilitator. so i would expect all but that row to be updated.

**Pull request merged**